Add username to onboarding

jacobherrington opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/master/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/master/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This change will allow a new user to change their username in onboarding.

## Related Tickets & Documents

Solves RFC: https://github.com/forem/rfcs/pull/47

## QA Instructions, Screenshots, Recordings

Onboarding with OAuth: https://www.loom.com/share/0c9e4dd792544264bfc11140d339bd04
Onboarding with Email: https://www.loom.com/share/7a64332c27244de4b345b09f7c3e0662

When the username is unavailable:
![image](https://user-images.githubusercontent.com/11466782/108103899-30010580-7050-11eb-8310-5d3509c49614.png)

When the username is invalid:
![image](https://user-images.githubusercontent.com/11466782/108104708-3f348300-7051-11eb-98a5-279ee59829d6.png)

There might be some opportunity to add front-end validation here, but the way we submit this form is a little odd so I'd have to play around with it a bit to figure out exactly how that should be done. For now, I think the backend validation is sufficient and the user doesn't lose anything from the form if the username fails validation so the UX isn't too bad imo.

## Added tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

citizen428 commented on app/controllers/users_controller.rb:
```diff
@@ -165,19 +165,21 @@ def remove_identity
     end
 
-    if current_user.save && params[:profile]
+    if current_user.invalid?
```
I know you're primarily updating code, but the `Profiles::Update` service is taking care of quite a few things that also happen here, like updating the `profile_updated_at` timestamp, etc. It also runs user and profile updates in a transaction, so it will either update both or neither.

It may be worth investigating if you could reduce this method to generating the correct user attributes and passing them into the service together with `profile_params`. IMHO the more we can standardize behavior the better.

jacobherrington replied:
That's a great suggestion. I believe this works as expected with only a minor tweak.

nickytonline commented on app/javascript/onboarding/components/ProfileForm.jsx:
```diff
@@ -50,23 +50,31 @@ export class ProfileForm extends Component {
+    const { username, ...newFormValues } = formValues;
     fetch('/onboarding_update', {
       method: 'PATCH',
       headers: {
         'X-CSRF-Token': csrfToken,
         'Content-Type': 'application/json',
       },
       body: JSON.stringify({
-        user: { last_onboarding_page },
-        profile: { ...formValues },
+        user: { last_onboarding_page, username },
+        profile: { ...newFormValues },
       }),
       credentials: 'same-origin',
```
refactor (blocking): Use the `request` utility function instead.

Since you're modifying the fetch call, might as well migrate it to our `request` function. It can be imported like so `import { request } from '@utilities/http';`.

It handles stringifying the payload and all the headers you need by default.

```suggestion
    request('/onboarding_update', {
      method: 'PATCH',
      body: {
        user: { last_onboarding_page, username },
        profile: { ...newFormValues },
      },
```

nickytonline commented on app/javascript/onboarding/components/ProfileForm.jsx:
```diff
@@ -50,23 +50,31 @@ export class ProfileForm extends Component {
-    }).then((response) => {
-      if (response.ok) {
+    })
+      .then((response) => {
```
refactor (non-blocking): Consider using `async`/`await`.

nickytonline commented on app/javascript/onboarding/components/ProfileForm.jsx:
```diff
@@ -219,9 +224,20 @@ export class ProfileForm extends Component {
             </div>
-            <div>{sections}</div>
+            <div>
+              <div className="onboarding-profile-sub-section">
```
refactor (question): Does this `<div />` need to be nested in a `<div />`?

nickytonline commented on app/javascript/onboarding/components/ProfileForm/TextInput.jsx:
```diff
@@ -26,7 +26,13 @@ import { FormField } from '@crayons';
 
 export function TextInput(props) {
   const { onFieldChange } = props;
-  const { attribute_name, placeholder_text, description, label } = props.field;
```
refactor (nitpick): `field` can be destructured from `props`.

```suggestion
  const { onFieldChange, field } = props;
  const {
    attribute_name,
    placeholder_text,
    default_value,
    description,
    label,
  } = field;
```

nickytonline reviewed (changes requested):
Thanks for the PR @jacobherrington! Some suggestions and a request change.

citizen428 reviewed (approved):
Approving from the Ruby side, pending fixes to Nicky's suggestions/concerns.

jacobherrington commented:
@nickytonline I think I addressed all of your comments, thanks for the suggestions!

nickytonline reviewed (commented):
Jacob, thanks for making the changes. I agree that adding some front-end validation would help. We could prevent the user from continuing by disabling the button until all fields are valid, but I think we would need to do this for the whole onboarding flow so the experience would be uniform, i.e. a separate PR. At least right now, there is backend validation.

A couple of things I noticed.

There is no validation if the username is left blank. From a user perspective, that looks like an error that wasn't caught. I know after by going through the flow it gives me a username which appears to be generated from the e-mail minus the domain and if from social media, the social media handle.

https://user-images.githubusercontent.com/833231/108450610-81b5b580-7233-11eb-93c4-fe7800806689.mp4

Another thing I noticed is the navigation button says "skip for now" when on the profile form even though there is a username filled out. It will only change to "continue" if you edit any of the fields. I'm not sure if this would confuse a user or not that we're skipping setting their username.

https://user-images.githubusercontent.com/833231/108450653-8ed2a480-7233-11eb-9f27-cfe65ccc5454.mp4

Lastly, we should definitely have end to end tests for the onboarding flow, but perhaps in a follow up PR since you've updated the component tests for your changes.

jacobherrington commented:
@nickytonline Great catch, submitting a blank username would be super confusing! I changed the ProfileForm component to "not be skippable" so the button always reads "Continue." To address the empty username case, I updated the endpoint to throw an error if we get a blank username param from the form (even though the app will handle a blank username, AFAIK).

![image](https://user-images.githubusercontent.com/11466782/108543580-e2cca000-72aa-11eb-8d7b-9bbe48e1f3c7.png)

I also added the required attribute to the field, which doesn't necessarily work as one might expect, but it does highlight the field red if it is blank when it loses focus (screenshot below) - so that should be a helpful short-term fix.

![image](https://user-images.githubusercontent.com/11466782/108543485-c16bb400-72aa-11eb-820e-795f3671e2ca.png)

We should definitely revisit front-end validation on the entire onboarding flow! Maybe we should try to put an issue into the next cycle for that task?

I think adding E2E tests here is also a great idea. That might be another great task for the next cycle (and I'd love to work on that, as an excuse to write some Cypress tests).

nickytonline commented:
Taking this for another spin @jacobherrington.

citizen428 commented on app/controllers/users_controller.rb:
```diff
@@ -169,6 +169,10 @@ def onboarding_update
     user_params = { saw_onboarding: true }
 
     if params[:user]
+      if params[:user].key?(:username) && params[:user][:username].blank?
```
Ruby added the `dig` method a while back to make checks like this one easier:

```suggestion
      if params.dig(:user, :username).blank?
```

jacobherrington replied:
I had to tweak a test because there was a case in which we sent:  `{ user: { last_onboarding_page: "v2: personal info form" } }` causing this check to raise an error. However, I don't think that test reflects the way this form is going to be used and we should always expect a username so that's fine.

nickytonline commented on app/javascript/onboarding/components/ProfileForm.jsx:
```diff
@@ -47,26 +47,27 @@ export class ProfileForm extends Component {
+        },
+      });
+      if (!response.ok) {
+        throw response;
```
Refactor (blocking): Provide feedback to the user if there is an issue with the response. Throwing the response here only log to the browser's console and there is no feedback to the user. Opt to set the `errorMessage` state property with a friendly generic message, e.g. "Unable to continue, an error occurred. Please try again." or something along those lines. I imagine we have wording for a message like this elsewhere in the application.

Note, to test this, I changed the request to give a 404.

<img width="1815" alt="image" src="https://user-images.githubusercontent.com/833231/108648356-ffbeca00-7488-11eb-8e97-ee3bfcaec007.png">

jacobherrington replied:
@nickytonline Addressed this in the last commit! The only error I'm sending custom messages for is 422, so anything else should be handled in this more generic way for the time being:
![image](https://user-images.githubusercontent.com/11466782/108892802-ed05db80-75d5-11eb-8ade-c180a327291b.png)

Good call on the Honeybadger logging - totally forgot to include that. :smile:

nickytonline replied:
I see the generic error visually now. 😎

refactor (blocking): Error appears saying username can't be blank when you click continue even if the username is populated.

![image](https://user-images.githubusercontent.com/833231/108901963-825e9b00-75e9-11eb-94f8-e7f49c574d80.png)

The issue appears to be that the username is not in the component state yet. It gets set when you change the username field.

![image](https://user-images.githubusercontent.com/833231/108902256-e2edd800-75e9-11eb-8738-ef0808f92cae.png)

You can set up the initial state like so in the constructor to resolve the issue.

```javascript
    this.state = {
      groups: [],
      formValues: { username: this.user.username },
      canSkip: false,
      last_onboarding_page: 'v2: personal info form',
    };
```

nickytonline reviewed (changes requested):
This is looking great @jacobherrington. One change request and I think this is good to go.

nickytonline reviewed (changes requested):
One more request change.

nickytonline commented on app/javascript/onboarding/components/ProfileForm.jsx:
```diff
@@ -22,7 +22,7 @@ export class ProfileForm extends Component {
     this.state = {
       groups: [],
-      formValues: {},
+      formValues: { user: { username: this.user.username } },
```
refactor (blocking): It needs to be just a `username` property in the component state, not `user.username`.

Below you can see the initial state in `user.username`, but when you change the username, it's populated under just `username` in the component state.

![image](https://user-images.githubusercontent.com/833231/108940129-d173f280-7620-11eb-8b17-70bccdac4210.png)

jacobherrington replied:
:upside_down_face:

jacobherrington commented:
Loom doesn't seem to work for me at the moment, but I recorded myself walking through this: 


https://user-images.githubusercontent.com/11466782/109028540-84773700-7687-11eb-9f36-bb45bed185fc.mp4


We should definitely try to get some Cypress tests in place over onboarding ASAP. :upside_down_face:

nickytonline reviewed (approved):
Thanks for making the changes @jacobherrington! 🚀
